employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/SkillService.java

```diff
@@ -85,21 +85,15 @@ public Skill updateSkill(Integer tenantId, Skill skill) {
 
         Optional<Skill> skillOptional = skillRepository.findById(skill.getId());
 
-        //If name is null
         if (skill.getName() == null) {
             throw new IllegalStateException("Skill entity name with tenantId (" + tenantId + ") cannot be null.");
-        }
-        //Skill not found
-        else if (!skillOptional.isPresent()) {
+        } else if (!skillOptional.isPresent()) {
             throw new EntityNotFoundException("Skill entity with ID (" + skill.getId() + ") not found.");
-        }
-        //Skill changes tenantId
-        else if (!skillOptional.get().getTenantId().equals(skill.getTenantId())) {
+        } else if (!skillOptional.get().getTenantId().equals(skill.getTenantId())) {
             throw new IllegalStateException("Skill entity with tenantId (" + skillOptional.get().getTenantId()
                     + ") cannot change tenants.");
         }
 
-        skillRepository.deleteById(skill.getId());
         return skillRepository.save(skill);
```

Comment:
`skill` instance here is the one coming from the request, it might have an old version or lacking an id. A better approach would be:
```java
skillEntity = reposiotry.findById(id); // or skillOptional.get() after ensuring it exists
skillEntity.setName(skill.getName());
... modify other fields
```
Pay closer attention to the memory model and lifecycle of jpa entities (attached/detached ...). It'll be more important when you deal with more complex classes, check how this is handled in [shift roster](https://github.com/kiegroup/optaweb-employee-rostering/blob/3df530af050b235f7ed8614e44db833c788b26b8/employee-rostering-server/src/main/java/org/optaweb/employeerostering/server/roster/RosterRestServiceImpl.java#L334).

Reply:
Yes, using `Skill` instance from request parameter will lead to optimistic locking exceptions when version is outdated. Is it necessary to save the repository `skillEntity` after modifying its fields? Point 5 in this tutorial goes over updating entities, and it uses `repository.save()` after modifying fields: https://www.baeldung.com/spring-data-crud-repository-save. Otherwise it's consistent with your approach.

I'll do more research into the JPA entity lifecycle and their states. Good catch!